alt_src/alt_src.py

```diff
@@ -1712,6 +1723,7 @@ def run(self):
         self.push_git(state)
         self.set_state('PUSHED')
         self.notify()
+        self.clean_lookaside_cache()
```

Comment:
I think it would be better for this to happen a lot earlier, probably before push_git (that won't break anything, will it?)

If this is done as the last step, then this seems possible:

- lookaside cache is populated
- push_git succeeds
- then we crash before we get to clean_lookaside_cache
- due to the crash, failed pub task is retried, but this time it detects there's nothing to be done (in duplicate_check)
- therefore lookaside cache is never cleaned until we happen to process same package & branch in the future (which could be a long time away, or never)

Thus the command is no longer idempotent, if interrupted part way through it may or may not clean if re-run.

Reply:
Looking at this a bit more, I see that there's an earlier call to clean_lookaside_cache (in Stager) as well.

I'm a bit confused about the relationship between them.  Usually, Stager is run followed by Pusher.  If Stager already did clean_lookaside_cache, does Pusher (push_lookaside method) still work correctly?

Reply:
> I think it would be better for this to happen a lot earlier, probably before push_git (that won't break anything, will it?)
> 
> If this is done as the last step, then this seems possible:
> 
> * lookaside cache is populated
> * push_git succeeds
> * then we crash before we get to clean_lookaside_cache
> * due to the crash, failed pub task is retried, but this time it detects there's nothing to be done (in duplicate_check)
> * therefore lookaside cache is never cleaned until we happen to process same package & branch in the future (which could be a long time away, or never)
> 

If clean_lookaside_cache is before push_git and at any point the push fails, the state remains as 'staged' and the push will be retried including push_lookaside. If the files were cleaned before the push completed, it will try to rsync again and will fail as the files for lookaside were already removed.

> Thus the command is no longer idempotent, if interrupted part way through it may or may not clean if re-run.

Agree, the command might leave some files if the push failed. But usually the script is used to push one package at a time and shouldn't leave much behind unless it's a series of failures during push to centos git.

> Looking at this a bit more, I see that there's an earlier call to clean_lookaside_cache (in Stager) as well.
>
> I'm a bit confused about the relationship between them. Usually, Stager is run followed by Pusher. If Stager already did > clean_lookaside_cache, does Pusher (push_lookaside method) still work correctly?

clean_lookaside_cache is called in Stager 'before' any content is staged to cleanup any files from previous push (if there are content to be staged in current push) . One in Pusher cleans up the files from the current push once the push completes.